src/python/py_generate_pipeline.cpp

```diff
@@ -263,30 +267,54 @@ py::object call_common_generate(
     LLMPipeline& pipe, 
     const std::variant<ov::Tensor, TokenizedInputs, std::string, std::vector<std::string>>& inputs, 
     const OptionalGenerationConfig& config, 
-    const StreamerVariant& streamer, 
+    const PyBindStreamerVariant& py_streamer, 
     const py::kwargs& kwargs
 ) {
     auto updated_config = update_config_from_kwargs(config, kwargs);
+    py::object results;
     EncodedInputs tensor_data;
-
-    if (auto data = std::get_if<ov::Tensor>(&inputs)) {
-        return py::cast(pipe.generate(*data, updated_config, streamer));
-    } else if (auto data = std::get_if<TokenizedInputs>(&inputs)) {
-        return py::cast(pipe.generate(*data, updated_config, streamer));
-    } else if (auto data = std::get_if<std::string>(&inputs)) {
-        DecodedResults res = pipe.generate(*data, updated_config, streamer);
+    StreamerVariant streamer = std::monostate();
+    
+    std::visit(overloaded {
+    [&streamer](const std::function<py::object(py::object)>& py_callback){
+        // Wrap python streamer with manual utf-8 decoding. Do not rely
+        // on pybind automatic decoding since it raises exceptions on incomplete strings.
+        auto callback_wrapped = [&py_callback](std::string subword) -> bool {
+            auto py_str = PyUnicode_DecodeUTF8(subword.data(), subword.length(), "replace");
+            return py_callback(py::reinterpret_borrow<py::object>(py_str)).cast<bool>();
+        };
+        streamer = callback_wrapped;
+    },
+    [&streamer](std::shared_ptr<StreamerBase> streamer_cls){
+        streamer = streamer_cls;
```

Comment:
It seems StreamerBase must also perform custom `PyUnicode_DecodeUTF8()`

Reply:
StreamerBase works only with encoded token ids, therefore there are no strings and utf problems.

I user wished to write his own streamer, inherits StreamerBase and detokenizes with our tokenizer then we will handle that
https://github.com/openvinotoolkit/openvino.genai/blob/master/src/python/py_generate_pipeline.cpp#L423-L427